packages/code-du-travail-data/dataset/datafiller/fix-urls.js

```diff
@@ -0,0 +1,94 @@
+const fetch = require("node-fetch");
+const pLimit = require("p-limit");
+
+const { getReference } = require("./elastic");
+
+/*
+fix datafiller urls based on ES results
+*/
+
+const DATAFILLER_URL =
+  process.env.DATAFILLER_URL || "https://datafiller.num.social.gouv.fr";
+
+const RECORDS_URL = `${DATAFILLER_URL}/kinto/v1/buckets/datasets/collections/requetes/records`;
+
+const serial = pLimit(1);
+
+const replaceReference = async (title, ref) => {
+  const hit = await getReference(title, ref);
+  if (hit) {
+    const oldUrl = ref.url;
+    const newUrl = ref.url.replace(/^(\/[^/]+\/)(.*)/, `$1${hit._source.slug}`);
+    if (newUrl !== oldUrl) {
+      console.log("FIX URL", oldUrl, newUrl);
+      return {
+        ...ref,
+        // replace slug only
+        url: newUrl
+      };
+    }
+  }
+  return ref;
+};
+
+const sortByKey = key => (a, b) => {
+  if (a[key] < b[key]) {
+    return 1;
+  } else if (a[key] > b[key]) {
+    return -1;
+  }
+  return 0;
+};
+
+// check refs against ES
+const getReferences = row =>
+  Promise.all(
+    row.refs.filter(ref => ref.url).map(ref => replaceReference(row.title, ref))
+  ).then(arr => arr.filter(Boolean).sort(sortByKey("relevance")));
+
+const fixAll = async () =>
+  fetch(RECORDS_URL)
+    .then(res => res.json())
+    .then(json => json.data)
+    // DEBUG
+    // .then(rows =>
+    //   rows.filter(row => row.title === "rupture-assistance-maternelle")
+    // )
+    .then(rows =>
+      Promise.all(
+        rows.map(row =>
+          serial(async () => ({
+            ...row,
+            refs: await getReferences(row)
+          }))
+        )
+      )
+    );
+// DEBUG
+//.then(rows => console.log(JSON.stringify(rows, null, 2)) || rows);
+
+if (require.main === module) {
```

Comment:
I wonder If we could have an updateDatafillerRef() that we can call in the fetch-requete that would limit the use of fix-url script

Reply:
or a "prestart" script ?

Reply:
my concern was more about the expensiveness that run requests on ES two times
but yup good idea too

Reply:
added a prestart script